src/main/java/com/zepben/testutils/exception/ExceptionMatcher.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2020 Zeppelin Bend Pty Ltd
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+package com.zepben.testutils.exception
+
+import kotlin.Throws
+import java.util.regex.Pattern
+
+class ExceptionMatcher<T : Throwable>(val exception: T) {
+
+    @Throws(ExpectExceptionError::class)
+    fun withMessage(expected: String): ExceptionMatcher<T> {
+        if (expected == exception.message) return this
+        throw ExpectExceptionError(expected, exception.message)
+    }
+
+    @Throws(ExpectExceptionError::class)
+    fun withMessage(expected: Pattern): ExceptionMatcher<T> {
+        val message = exception.message ?: throw ExpectExceptionError(expected.pattern(), "")
+        if (expected.matcher(message).matches()) return this
+        throw ExpectExceptionError(expected.pattern(), message)
+    }
+
+    @Throws(ExpectExceptionError::class)
+    fun withoutMessage(): ExceptionMatcher<T> {
+        if (exception.message == null || exception.message == "") return this
+        throw ExpectExceptionError("", exception.message)
+    }
+
+    @Throws(ExpectExceptionError::class)
+    inline fun <reified R : Throwable> withCause(): ExceptionMatcher<T> {
+        val actual = exception.cause ?: throw ExpectExceptionError(R::class.simpleName, "")
+        if (actual is R) return this
+        throw ExpectExceptionError(R::class.simpleName, actual.javaClass.simpleName)
+    }
+
+    /**
+     * Java-interoperable version of withCause()
+     */
+    @Throws(ExpectExceptionError::class)
+    fun <R : Throwable> withCause(clazz: Class<R>): ExceptionMatcher<T> {
+        val actual = exception.cause ?: throw ExpectExceptionError(clazz.simpleName, "")
+        if (clazz.isInstance(actual)) return this
+        throw ExpectExceptionError(clazz.simpleName, actual.javaClass.simpleName)
+    }
+
+    @Throws(ExpectExceptionError::class)
+    fun <R: Throwable> withAnyCause(): ExceptionMatcher<T> = withCause<Throwable>()
```

Comment:
No coverage of this function.

Nit: Missing space between `R:`. File formatting (default CTRL+ALT+L) should fix this.